{{cookiecutter.project_slug}}/.github/workflows/test-pr.yml

```diff
@@ -0,0 +1,85 @@
+name: 'Run Tests'
+on:
+  pull_request:
+  workflow_dispatch:
+concurrency:
+  group: {% raw %}${{ github.workflow }}-${{ github.ref }}{% endraw %}
+  cancel-in-progress: true
+
+jobs:
+  code-quality-checks:
+    name: 'Code Quality Checks'
+    runs-on: ubuntu-latest
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v3
+      - name: 'Install dependencies'
+        run: |
+          curl -sSL https://install.python-poetry.org | sudo bash -c 'POETRY_HOME=/usr python3 -'
+          poetry --version
+      - name: 'Build and run code quality checks'
+        run: make check
+
+  all-tests:
+    needs: code-quality-checks
+    name: 'All tests'
+    runs-on: [self-hosted, linux, normal]
+    strategy:
+      matrix:
+        distro: ['focal', 'jammy']
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v3
+      - name: 'Build Docker image'
+        run: |
+          COMMIT=$(git rev-parse --short=7 HEAD)
+          TAG={% raw %}${{ matrix.os }}{% endraw %}
+          CONTAINER={{ cookiecutter.project_slug }}-ci
+          IMAGE=runtimeverificationinc/${CONTAINER}:${TAG}
+
+          docker build .           \
+            --build-arg TAG=${TAG} \
+            --tag ${IMAGE}
+
+          docker run                 \
+            --name ${CONTAINER}      \
+            --rm                     \
+            --interactive            \
+            --tty                    \
+            --detach                 \
+            ${IMAGE}
+
+          docker cp . ${CONTAINER}:/home/user
+          docker exec ${CONTAINER} chown -R user:user /home/user
```

Comment:
We should pull this out into an action, so that it's easier for users to set up docker for mulitple different stages if their project calls for that. Here is an [example of a similar action](https://github.com/runtimeverification/k/blob/master/.github/actions/with-docker/action.yml), and [how it is used here](https://github.com/runtimeverification/k/blob/master/.github/workflows/test-pr.yml#L85)

If you want to leave this for a later PR that is OK too, but in terms of starting out repos from a better starting point I think this will help.

Freeman and I are still working on getting rid of the need for `docker exec ...` prefixing each test line too.

Reply:
Also, @F-WRunTime look how here we are using `docker cp . ${CONTAINER}:/home/user` followed by `docker exec ${CONTAINER} chown -R user:user /home/user`. I think this could be useful other places, instead of the volume mounting technique we're doing.

Benefit is that you know your environment will be cleaner each time, and the copied in data is not persistent.

Drawback is that the copied in data is not persistent, so pushing artifacts around between builds  is probably harder.

Reply:
As well as publishing anything from a workflow. BUT as we've been using nix and all that publishing artifacts via workflow artifact pushes to github has become less so necessary. As well as introducing pypi. 

SO I agree I don't have a hardline at this time on mounts. As we've been moving further from need to publish artifacts the host runner & action needs read/write access to to push. 
This did  however introduce permission issues still with copying files into a container that had no knowledge of the user that copied them in. 
HOWEVER Combine the cp method with our method of building containers with the provided knowledge of the running host user UID/GID into the container, I think we hit a sweet spot of proper user UID information and internal container clean-state. 

Let me know if that was too much ramble of current progress we've made with these container builds. 

**TLDR** I agree, lets make sure we build the containers and create this basic 'user'  with the UID/GID of our non-permissive executing 'github-runner' from using `id -u` and `id -g` and we're in a good place to avoid obscure permission issues.